File: backend/models/Lobby.js
const Deck = require("../card_management/cards");
const Player = require("./Player");
const Round = require("./Round");
const Statuses = require("./Statuses");

module.exports = class Lobby {
    constructor(lobbyId, playerId, nickname, roundDuration, maxRoundNumber) {
        this.id = lobbyId;
        this.status = Statuses.open;
        this.roundDuration = roundDuration;
        this.maxRoundNumber = maxRoundNumber;
        this.currentRound = 0;
        this.round = new Round()
        this.deck = new Deck()
        this.playerList = new Map([[playerId, new Player(playerId, nickname)]]);
        this.playerList.get(playerId).isMyTurn=true;
        this.round.currentBlackCard=this.deck.drawBlackCard;
    }
    
    getOpenLobbyData() {
        if (this.status != Statuses.open) throw 409
        const mappedPlayerList = Array.from(this.playerList.entries(),
            ([id, player]) => ({
                id,
                nickname: player.nickname,
                ready: player.isReady,
            }));
        return {
            "status": this.status,
            "roundDuration": this.roundDuration,
            "maxRoundNumber": this.maxRoundNumber,
            "players": mappedPlayerList,
        }
    }

    getOngoingLobbyData() {
        if (this.status == Statuses.open) throw 409
        const mappedPlayerList = Array.from(this.playerList.entries(),
            ([id, player]) => ({
                id,
                nickname: player.nickname,
                ready: player.isReady,
                score: player.score,
                isMyTurn: player.isMyTurn,
            }));
        return {
            "status": this.status,
            "roundDuration": this.roundDuration,
            "maxRoundNumber": this.maxRoundNumber,
            "currentRound":this.currentRound,
            "currentBlackCard": this.round.currentBlackCard,
            "players": mappedPlayerList,
        }
    }

    addPlayer(playerId, nickname){
        this.playerList.set(playerId, new Player(playerId, nickname));
    }

    setPlayerReady(playerId){
        const player = this.playerList.get(playerId);
        if(player == undefined) throw 404
        player.isReady=true;
    }

    resetAllPlayerReady(){
        this.playerList.forEach(p => {
            if(!p.isMyTurn)
            p.isReady = false;
        });
    }

    checkAllReady(){
        let everyOneReady = true;
        this.playerList.forEach(p => {
            if(!p.isReady) everyOneReady = false;
        });
        return everyOneReady;
    }

    playCard(playerId, cardIds){
        if (this.status != Statuses.play) throw 409
        const player = this.playerList.get(playerId);
        if(player == undefined) throw 404
        
        let playedCards = [];
        let found=0;
        let remainingCards = [];
        player.hand.forEach(c=>{
            if(cardIds.includes(c.id)){
                playedCards.push(c);
                found++;
            } else
                remainingCards.push(c); //todo
        });
        if(found != cardIds.length) throw 404
        const record = {
            "playerId":playerId,
            "playedCards":playedCards,
        }
        player.hand = remainingCards;
        this.round.playedCards.push(record);
    }

    nextBlackCard(){
        this.currentBlackCard=this.deck.drawBlackCard;
    }

    refillHands(){
        const max = 10;
        
        this.playerList.forEach((p)=>{
            let i;
            const missing= max-p.hand.length;
            for(i=0; i<missing; i++)
                p.hand.push(this.deck.drawWhiteCard); 
        })
    }
}

File: cah.server/src/main/java/com/touchgrass/cah/server/model/Lobby.java
package com.touchgrass.cah.server.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.touchgrass.cah.server.configuration.Constants;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Optional;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@ToString
public class Lobby {
    private String id;
    private LobbyStatus status;
    private int roundDuration;
    private int maxRoundNumber;
    private int currentRound;
    private Round round;
    @JsonIgnore
    private Deck deck;
    private HashMap<String, Player> playerList;

    public Lobby(int roundDuration, int maxRoundNumber, Player host, Deck deck) {
        id = "11111"; // UUID.randomUUID().toString();
        status = LobbyStatus.open;
        this.roundDuration = roundDuration;
        this.maxRoundNumber = maxRoundNumber;
        this.currentRound = 1;
        round = new Round();
        playerList = new HashMap<>();
        playerList.put(host.getId(), host);
        host.setMyTurn(true);
        this.deck = deck;
        round.setCurrentBlackCard(deck.drawBlackCard());
    }

    public void addPlayer(Player player) {
        this.playerList.put(player.getId(), player);
        System.out.println("Added Player " + player.getId() + " " + player.getNickname());
    }

    public void setPlayerReady(String playerId) throws CustomException {
        Player p = playerList.get(playerId);
        if (p == null) throw new CustomException(404, "Player not found");
        p.setReady(true);
        System.out.println("Set Player ready " + playerId + " " + this.playerList.get(playerId));
    }

    public boolean checkAllReady() {
        for (Player p : this.playerList.values()) {
            if (!p.isReady()) return false;

        }
        return true;
    }

    public void resetAllPlayerReady() {
        for (Player p : this.playerList.values()) {
            if (!p.isMyTurn()) { // necessary otherwise you'd need to set this manually in a less efficient manner
                p.setReady(false);
            }
        }
    }

    public void refillHands() {
        for (Player p : this.playerList.values()) {
            ArrayList<WhiteCard> hand = p.getHand();
            while (hand.size() < Constants.HAND_SIZE) {
                hand.add(deck.drawWhiteCard());
            }
        }
    }

    public void playCard(@Size(min = 1, max = 32) String playerId, @NotNull @Size(min = 1) List<@NotNull Integer> cardIds) throws CustomException {
        if (this.status != LobbyStatus.play) throw new CustomException(409, "Cannot play a card right now");
        Player player = this.playerList.get(playerId);
        if (player == null) throw new CustomException(404, "Player not found");

        ArrayList<WhiteCard> playedCards = new ArrayList<>();
        int found = 0;
        ArrayList<WhiteCard> remainingCards = new ArrayList<>();
        for (WhiteCard c : player.getHand()) {
            if (cardIds.contains(c.getId())) {
                playedCards.add(c);
                found++;
            } else
                remainingCards.add(c);
        }
        if (found != cardIds.size()) throw new CustomException(404, "Player possesses no such card");
        round.getPlayedCards().put(getNicknameFromId(playerId), playedCards);
        player.setHand(remainingCards);
    }

    public Player getTurnHolder() { // todo can optimize
        for (Player p : playerList.values()) {
            if (p.isMyTurn()) return p;
        }
        return null;
    }

    public void nextBlackCard() {
        round.setCurrentBlackCard(deck.drawBlackCard());
        round.setPlayedCards(new HashMap<>());
    }

    private String getNicknameFromId(String playerId) throws CustomException {
        Player player = playerList.get(playerId);
        if (player == null) throw new CustomException(404, "Player not found");
        return player.getNickname();
    }

    public Player playerFromNickname(String nickname) {
        Optional<Player> player = playerList.values().stream().filter(p -> p.getNickname().equals(nickname)).findFirst();
        return player.orElse(null);
    }
}
